File: L11-SieveofEratosthenes/CountNonDivisible.java
// N = [1...50000]
// A[i] = [1...2*N]

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class CountNonDivisible {
    // time complexity: O(N * log(N))
    // space complexity: O(N)
    // result: https://app.codility.com/demo/results/training7V9TWC-Z65/
    public static int[] solution(int[] A) {
        int N = A.length;
        if (N == 1) return new int[]{0};
        int[] ans = new int[N];
        Map<Integer, Integer> counters = new HashMap<>();
        for (int value : A) {
            if (counters.containsKey(value)) {
                counters.replace(value, counters.get(value) + 1);
            } else {
                counters.put(value, 1);
            }
        }
        for (int i = 0; i < N; i++) {
            int divisorCount = 0;
            for (int j = 1; j * j <= A[i]; j++) {
                if (A[i] % j == 0) {
                    int other = A[i] / j;
                    divisorCount += counters.containsKey(j) ? counters.get(j) : 0;
                    if (other != j) {
                        divisorCount += counters.containsKey(other) ? counters.get(other) : 0;
                    }
                }
            }
            ans[i] = N - divisorCount;
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println("Expect: [2, 4, 3, 2, 0]. Received: " + Arrays.toString(solution(new int[]{3, 1, 2, 3, 6})));
    }
}

File: L11-SieveofEratosthenes/CountNonDivisible.swift
import Foundation
// result: https://app.codility.com/demo/results/trainingSXYGJB-8G7/
public class CountNonDivisible {
    public func solution(_ A : inout [Int]) -> [Int] {
        let n = A.count
        if n == 1 { return [0] }
        var counter = [Int:Int]()
        for number in A {
            counter[number] = (counter[number] ?? 0) + 1
        }
        var answer = [Int]()
        for number in A {
            var a = 1
            // find divisors of A[i]
            var divisorCount = 0
            while a * a <= number {
                if number % a == 0 {
                    divisorCount += (counter[a] ?? 0) + (a * a != number ? counter[number / a] ?? 0 : 0)
                }
                a += 1
            }
            answer.append(n - divisorCount)
        }
        return answer
    }
}
